cognite/client/data_classes/data_modeling/instances.py

```diff
@@ -1446,3 +1448,305 @@ def __delitem__(self, key: str | tuple[str, str] | tuple[str, str, str]) -> None
+
+
+class TypedEdgeApply(EdgeApply, TypedInstance):
+    def __init__(
+        self,
+        space: str,
+        external_id: str,
+        type: DirectRelationReference | tuple[str, str],
+        start_node: DirectRelationReference | tuple[str, str],
+        end_node: DirectRelationReference | tuple[str, str],
+        existing_version: int | None = None,
+    ) -> None:
+        super().__init__(space, external_id, type, start_node, end_node, existing_version, None)
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
+
+    @classmethod
+    def _load(cls, resource: dict, cognite_client: CogniteClient | None = None) -> Self:
+        sources = resource.pop("sources", [])
+        properties = sources[0].get("properties", {}) if sources else {}
+        return cls._load_instance(resource, properties)
+
+    @property
+    def sources(self) -> list[NodeOrEdgeData]:
+        return [NodeOrEdgeData(source=self.get_source(), properties=self._dump_properties())]
+
+
+class TypedNode(Node, TypedInstance):
+    def __init__(
+        self,
+        space: str,
+        external_id: str,
+        version: int,
+        last_updated_time: int,
+        created_time: int,
+        deleted_time: int | None,
+        type: DirectRelationReference | tuple[str, str] | None,
+    ) -> None:
+        super().__init__(space, external_id, version, last_updated_time, created_time, deleted_time, None, type)
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
+
+    @classmethod
+    def _load(cls, resource: dict[str, Any], cognite_client: CogniteClient | None = None) -> Self:
+        all_properties = resource.pop("properties", {})
+        source = cls.get_source()
+        properties = all_properties.get(source.space, {}).get(source.as_source_identifier(), {})
+        return cls._load_instance(resource, properties)
+
+    @property
+    def properties(self) -> Properties:
+        return Properties({self.get_source(): self._dump_properties()})
+
+
+class TypedEdge(Edge, TypedInstance):
+    def __init__(
+        self,
+        space: str,
+        external_id: str,
+        version: int,
+        type: DirectRelationReference | tuple[str, str],
+        last_updated_time: int,
+        created_time: int,
+        start_node: DirectRelationReference | tuple[str, str],
+        end_node: DirectRelationReference | tuple[str, str],
+        deleted_time: int | None,
+    ) -> None:
+        super().__init__(
+            space,
+            external_id,
+            version,
+            type,
+            last_updated_time,
+            created_time,
+            start_node,
+            end_node,
+            deleted_time,
+            properties=None,
+        )
+
+    @staticmethod
+    @lru_cache(1)
+    def _base_properties() -> set[str]:
+        return _get_constructor_params(TypedEdge)
```

Comment:
You already cache `_get_constructor_params` btw

Reply:
Addressed 👍